From 9b9398567258fed12dfb22d0bc352545f33d54a0 Mon Sep 17 00:00:00 2001
From: Paul Thomson <paulthomson@google.com>
Date: Fri, 17 Apr 2026 12:50:51 +0100
Subject: [PATCH 1/8] KT-35614: getOrPut -> getOrPutNullable

Replace getOrPut -> getOrPutNullable. Without this, we are not caching a
null result for when a class does not exist, leading to large slowdown
if KotlinCliJavaFileManagerImpl#findClass is often called for a class
that does not exist.

https://youtrack.jetbrains.com/issue/KT-35614
---
 .../kotlin/cli/jvm/compiler/KotlinCliJavaFileManagerImpl.kt   | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/compiler/cli/cli-base/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinCliJavaFileManagerImpl.kt b/compiler/cli/cli-base/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinCliJavaFileManagerImpl.kt
index 810d4e612694..12a9c3376c5f 100644
--- a/compiler/cli/cli-base/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinCliJavaFileManagerImpl.kt
+++ b/compiler/cli/cli-base/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinCliJavaFileManagerImpl.kt
@@ -46,6 +46,7 @@ import org.jetbrains.kotlin.util.PhaseSideType
 import org.jetbrains.kotlin.util.tryMeasureSideTime
 import org.jetbrains.kotlin.utils.SmartList
 import org.jetbrains.kotlin.utils.addIfNotNull
+import org.jetbrains.kotlin.utils.getOrPutNullable
 
 // TODO: do not inherit from CoreJavaFileManager to avoid accidental usage of its methods which do not use caches/indices
 // Currently, the only relevant usage of this class as CoreJavaFileManager is at CoreJavaDirectoryService.getPackage,
@@ -94,7 +95,8 @@ class KotlinCliJavaFileManagerImpl(private val myPsiManager: PsiManager) : CoreJ
     private fun findVirtualFileForTopLevelClass(classId: ClassId, searchScope: GlobalSearchScope): VirtualFile? {
         val relativeClassName = classId.relativeClassName.asString()
         val outerMostClassFqName = classId.packageFqName.child(classId.relativeClassName.pathSegments().first())
-        return topLevelClassesCache.getOrPut(outerMostClassFqName) {
+        // KT-35614: getOrPut -> getOrPutNullable
+        return topLevelClassesCache.getOrPutNullable(outerMostClassFqName) {
             // Search java sources first. For build tools, it makes sense to build new files passing all the
             // class files for the previous build on the class path.
             //
-- 
2.54.0.1064.gd145956f57-goog

